src/chainstate/stacks/miner.rs

```diff
@@ -326,6 +393,218 @@ impl StacksBlockBuilder {
 
         test_debug!("\n\nMiner {}: Finished mining child of {}/{}. Trie is in mined_blocks table.\n", self.miner_id, self.chain_tip.burn_header_hash, self.chain_tip.anchored_header.block_hash());
     }
+    
+    /// Unconditionally build an anchored block from a list of transactions.
+    /// Used when we are re-building a valid block after we exceed budget
+    fn make_anchored_block_from_txs(mut builder: StacksBlockBuilder, chainstate: &mut StacksChainState, mut txs: Vec<StacksTransaction>) -> Result<StacksBlock, Error> {
+        debug!("Build anchored block from {} transactions", txs.len());
+        let mut epoch_tx = builder.epoch_begin(chainstate)?;
+        for tx in txs.drain(..) {
+            builder.try_mine_tx(&mut epoch_tx, &tx)?;
+        }
+        let block = builder.mine_anchored_block(&mut epoch_tx);
+        builder.epoch_finish(epoch_tx);
+        Ok(block)
+    }
+
+    /// Create a block builder for mining
+    fn make_block_builder(burndb: &mut BurnDB,
+                          chainstate: &mut StacksChainState,
+                          parent_burn_header_hash: &BurnchainHeaderHash,
+                          parent_block_hash: &BlockHeaderHash,
+                          proof: VRFProof,
+                          pubkey_hash: Hash160) -> Result<StacksBlockBuilder, Error> {
+
+        let builder = 
+            if *parent_burn_header_hash == FIRST_BURNCHAIN_BLOCK_HASH && *parent_block_hash == FIRST_STACKS_BLOCK_HASH {
+                StacksBlockBuilder::first_pubkey_hash(0, parent_burn_header_hash, FIRST_BURNCHAIN_BLOCK_TIMESTAMP, &proof, pubkey_hash)
+            }
+            else {
+                // building off an existing stacks block
+                let parent_stacks_block_snapshot = {
```

Comment:
Can this be supplied as a parameter, so that the block builder doesn't need a BurnDB instance?

Reply:
The thinking here is that eventually, `build_anchored_block` will be called via the RPC interface.  So if we don't load the snapshot here, the HTTP server will need to do so when it calls `build_anchored_block`.

But, would it make it easier for the neon node's development to just take the snapshot as an argument?

Reply:
> But, would it make it easier for the neon node's development to just take the snapshot as an argument?

For now, yes (and when we implement the HTTP endpoint, the HTTP server can probably handle loading and supplying the snapshot).

Reply:
I refactored the `build_anchored_block` method to work with the data available to the neon node.